<style>img {width: 700px;}</style>



<img src="./images/title.png"/>




<img src="./images/prob_1/p1.png"/>

<img src="./images/prob_1/a.png"/>

Dual problem presented in problem 1 :

minimize   $12\pi_{1} + 8\pi_{2}$

subject to : 

$x_{1}: 4\pi_{1} + \pi_{2} \geq 1 \\
x_{2}: \pi_{1} + 2\pi_{2} \geq 2 \\
x_{3}: 3\pi_{1} - 2\pi_{2} \geq -8$

### Feasable region plot
<img src="./images/fig1.png" style="width:460px;"/>

Figure 1


### Feasable solutions
<img src="./images/fig2.png" style="width:460px;"/>

Figure 2


As we can see in Figure 2, all feasable solutions lie in the feasable region's edge $x_{2}: \pi_{1} + 2\pi_{2} = 2$. Such line segment has vertex $(0,1)$, which is optimal. It has parametric equations:

$(\pi_{1}^{*},\pi_{2}^{*}) = (0,1) + \lambda$



<img src="./images/prob_1/b.png"/>

#### Optimality condition 1: 
$\bf{x^{*}}$ is primal feasable as it satisfies the inequalities in the primal problem

#### Optimality condition 2:
$\bf{\pi^{*}}$ is dual feasable as it satisfies the inequalities in the dual problem

#### Optimality condition 3:
Complementary slackness:

$x_{3}: 3\pi_{1} - 2\pi_{2} \geq -8$ always holds

- therefore: $x_{3} = 0$

Then we have:

$4x_{1} + x_{2} = 12 \\
x_{1} + 2x_{2} = 8 \\
x_{1} \geq 0, x_{2} \geq 0$

Solving for $x_{1}$ and $x_{2}$:

$x_{2} = 12 - 4x_{1} \\
x_{1} - 8x_{1} + 24 = 8 \\
-7x_{1} = -16$

Which results in:

$x_{1} = \frac{16}{7} \\
x_{2} = \frac{20}{7}$

 - therefore ${\bf x^{*}} = (\frac{16}{7}, \frac{20}{7}, 0)$




<img src="./images/prob_1/c.png"/>

The reduced costs of the primal variables are computed as following:

$\bar{c_{1}} = 4\pi_{1}^{*} + \pi_{2}^{*} - 1 = 4\lambda + 1 - \frac{\lambda}{2} - 1 = \frac{7\lambda}{2}  \\
\bar{c_{2}} = \pi_{1}^{*} + 2\pi_{2}^{*} - 2 = \lambda + 2 - \lambda - 2 = 0 \\
\bar{c_{3}} = 3\pi_{1}^{*} - 2\pi_{2}^{*} + 8 = 3\lambda - 2 - \lambda + 8 = 2\lambda + 6$

Where $\bf{\pi^{*}}$ is a dual optimal solution
 - therefore, the reduced costs are not unique

<img src="./images/prob_1/d.png"/>

primal problem:

maximize $x_{1} + 2x_{2} - 8x_{3}$

subject to :

$4x_{1} + x_{2} + 3x_{3} = 12 \\
x_{1} + 2x_{2} - 2x_{3} = 8 \\
x_{1} \geq 0, x_{2} \geq 0, x_{3} \geq 0$

Let's consider $\bf{\pi^{*}} = (0,1)$ we have $x_{3} = 0$ and then:

$4x_{1} + x_{2} = 12 + \Delta b_{1} \\
x_{1} + 2x_{2} = 8 + \Delta b_{2}$

Solving we get:

$\hat{x}^{*} = (\frac{1}{7} (2\Delta b_{1} - \Delta b_{2} + 16), \frac{1}{7} (-\Delta b_{1} + 4\Delta b_{2} + 20), 0)$

Imposing primal feasability returns:

$2\Delta b_{1} - \Delta b_{2} \geq -16 \\
-\Delta b_{1} + 4\Delta b_{2} \geq -20$

As a result: if $\Delta b_{1} = 0$ then $\Delta b_{2} \in [-5, 16]$ and if $\Delta b_{2} = 0$ then $\Delta b_{1} \in [-8, 20]$



<img src="./images/prob_1/e.png"/>

The primal problem only has one feasable solution, so the optimal solution $\bf{x^{*}}$ remains optimal for the following modified problem:

maximize $(1 + \Delta r_{1})x_{1} + (2 + \Delta r_{2})x_{2} - (8 + \Delta r_{3})x_{3}$

subject to :

$4x_{1} + x_{2} + 3x_{3} = 12 \\
x_{1} + 2x_{2} - 2x_{3} = 8 \\
x_{1} \geq 0, x_{2} \geq 0, x_{3} \geq 0$

Following a standard analysis, we consider ${\bf\hat{x}^{*}} = (\frac{16}{7}, \frac{20}{7}, 0)$

By complementary slackness we formulate the following 2 systems of linear equations:

#### system 1:

$4\pi_{1} + \pi_{2} = 1 + \Delta r_{1} \\
\pi_{1} + 2\pi_{2} = 2 + \Delta r_{2}$

resulting in: 

$\hat{\pi}^{*} = (\frac{1}{7} (2\Delta r_{1} - \Delta r_{2}), \frac{1}{7} (7 - \Delta r_{1} + 4\Delta b_{2}))$

Dual feasability means that:


$3\hat{\pi}^{*}_{1} + 2\hat{\pi}^{*}_{2} + 8 - \Delta r_{3} = \frac{1}{7} (42 + 8\Delta r_{1} - 11\Delta r_{2} - 7\Delta r_{3})$

and:

$8\Delta r_{1} - 11\Delta r_{2} - 7\Delta r_{3} \geq -42$

So, if:

$\Delta r_{1} = 0, \Delta r_{2} = 0$, then $\Delta r_{3} \leq 6$

$\Delta r_{1} = 0, \Delta r_{3} = 0$, then $\Delta r_{2} \leq \frac{42}{11}$

$\Delta r_{2} = 0, \Delta r_{3} = 0$, then $\Delta r_{1} \geq \frac{-21}{4}$

#### system 2:

$\pi_{1} + 2\pi_{2} = 2 + \Delta r_{2} \\
3\pi_{1} - 2\pi_{2} = -8 + \Delta r_{3}$

resulting in: 

$\hat{\pi}^{*} = (\frac{1}{4} (- 14 - 3\Delta r_{2} + \Delta r_{3}), \frac{1}{4} (- 20 -2\frac{1}{2}\Delta r_{2} + 2\frac{1}{2}\Delta r_{3}))$

Dual feasability means that:


$4\hat{\pi}^{*}_{1} + \hat{\pi}^{*}_{2} -1 - \Delta r_{1} = \frac{1}{4} (- 98 - 4\Delta r_{1} - 11\Delta r_{2} + 9\Delta r_{3})$

and:

$-4\Delta r_{1} - 11\Delta r_{2} + 9\Delta r_{3} \geq 98$

So, if:

$\Delta r_{1} = 0, \Delta r_{2} = 0$, then $\Delta r_{3} \geq \frac{98}{9}$

$\Delta r_{1} = 0, \Delta r_{3} = 0$, then $\Delta r_{2} \leq \frac{98}{11}$

$\Delta r_{2} = 0, \Delta r_{3} = 0$, then $\Delta r_{1} \leq \frac{-49}{2}$

Combining both cases we obtain that $\bf x^{*}$ is optimal for any $\Delta r_{j} \in \mathbb{R}$

<img src="./images/prob_2/p2.png"/>

<img src="./images/prob_2/a.png"/>

The primal LO formulation for LR with MAE criterion as seen in class has the following parameters:

- Objective: minimize $\sum_{i=1}^{n} (e_{i}^{+} + e_{i}^{-})$

- Subject to the following constraints: 
    - $(e_{i}^{+} - e_{i}^{-}) + b_{0} + x_{i1} b_{1} + \dots + x_{im} b_{m} = y_{i}$ $i = 1, \dots, n$

    - $e_{i}^{+}, e_{i}^{-} \geq 0,$ $i = 1, \dots, n$

In [8]:
# importing libraries
import gurobipy as gp
from gurobipy import GRB

In [9]:
# creating multidict with observations
observations, x, y = gp.multidict({
    ('1'): [30.1,176.2],
    ('2'): [29.5,176.8],
    ('3'): [30.4,184.2],
    ('4'): [31.6,173.2],
    ('5'): [27.4,172.8],
    ('6'): [28.3,174.1],
    ('7'): [33.4,180.5],
})

In [10]:
model = gp.Model('CurveFitting')

# Constant term of the function f(x). This is a free continuous variable that can take positive and negative values. 
a = model.addVar(lb=-GRB.INFINITY, ub=GRB.INFINITY, vtype=GRB.CONTINUOUS, name="a")

# Coefficient of the linear term of the function f(x). This is a free continuous variable that can take positive 
# and negative values.
b = model.addVar(lb=-GRB.INFINITY, ub=GRB.INFINITY, vtype=GRB.CONTINUOUS, name="b")

# Non-negative continuous variables that capture the positive deviations
e_plus = model.addVars(observations, vtype=GRB.CONTINUOUS, name="e_plus")

# Non-negative continuous variables that capture the negative deviations
e_neg = model.addVars(observations, vtype=GRB.CONTINUOUS, name="e_neg")

# Non-negative continuous variables that capture the value of the maximum deviation
z = model.addVar(vtype=GRB.CONTINUOUS, name="z")

In [11]:
# Deviation constraints

deviations = model.addConstrs( (b*x[i] + a + e_plus[i] - e_neg[i] == y[i] for i in observations), name='deviations')

# Objective function of problem 1

model.setObjective(e_plus.sum('*') + e_neg.sum('*'))

# Verify model formulation

model.write('CurveFitting.lp')

# Run optimization engine

model.optimize()

Gurobi Optimizer version 9.0.3 build v9.0.3rc0 (linux64)
Optimize a model with 7 rows, 17 columns and 28 nonzeros
Model fingerprint: 0xc72524f5
Coefficient statistics:
  Matrix range     [1e+00, 3e+01]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+02, 2e+02]
Presolve removed 0 rows and 1 columns
Presolve time: 0.00s
Presolved: 7 rows, 16 columns, 28 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0      handle free variables                          0s
       7    1.4029412e+01   0.000000e+00   0.000000e+00      0s

Solved in 7 iterations and 0.00 seconds
Optimal objective  1.402941176e+01


In [12]:
# The best straight line that minimizes the absolute value of the deviations is:
print(f"y = {b.x:.4f}x + ({a.x:.4f})")

y = 1.2549x + (138.5863)


In [13]:
# Maximum deviation constraints

maxPositive_deviation = model.addConstrs( (z >= e_plus[i] for i in observations), name='maxPositive_deviation')

maxNegative_deviation = model.addConstrs( (z >= e_neg[i] for i in observations), name='maxNegative_deviation')

# Objective function for Problem 2

model.setObjective(z)

# Run optimization engine

model.optimize()

Gurobi Optimizer version 9.0.3 build v9.0.3rc0 (linux64)
Optimize a model with 21 rows, 17 columns and 56 nonzeros
Coefficient statistics:
  Matrix range     [1e+00, 3e+01]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+02, 2e+02]
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.402941e+01   0.000000e+00      0s
       4    5.5571429e+00   0.000000e+00   0.000000e+00      0s

Solved in 4 iterations and 0.00 seconds
Optimal objective  5.557142857e+00


In [14]:
# The best straight line that minimizes the maximum deviation is: 
print(f"y = {b.x:.4f}x + ({a.x:.4f})")

y = 0.0952x + (175.7476)


<img src="./images/prob_2/b.png"/>

The prediction equation (representing the optimal solution) resulting from the implementation is $\hat{y} = 1.2549x + 138.5863$. 